File: BlogApp-RN/src/core/errors/handleException.ts
import {
  AuthException,
  CacheException,
  DataFormatException,
  FileSystemException,
  ServerException,
  UnknownException,
} from "./exceptions";
import {
  AuthFailure,
  CacheFailure,
  DataFormatFailure,
  Failure,
  FileSystemFailure,
  ServerFailure,
  UnknownFailure,
} from "./failures";

export const handleException: (exception: unknown) => Failure = (exception) => {
  if (exception instanceof AuthException)
    return AuthFailure.fromException(exception);
  if (exception instanceof ServerException)
    return ServerFailure.fromException(exception);
  if (exception instanceof FileSystemException)
    return FileSystemFailure.fromException(exception);
  if (exception instanceof DataFormatException)
    return DataFormatFailure.fromException(exception);
  if (exception instanceof CacheException)
    return CacheFailure.fromException(exception);
  if (exception instanceof UnknownException)
    return UnknownFailure.fromException(exception);
  return new UnknownFailure("An uknown error has occurred", 500);
};

File: blogapp_flutter/lib/core/errors/handle_exception.dart
import 'exceptions.dart';
import 'failures.dart';

Failure handleException(Object exception) {
  if (exception is AuthException) {
    return AuthFailure.fromException(exception);
  } else if (exception is ServerException) {
    return ServerFailure.fromException(exception);
  } else if (exception is FileSystemException) {
    return FileSystemFailure.fromException(exception);
  } else if (exception is DataFormatException) {
    return DataFormatFailure.fromException(exception);
  } else if (exception is CacheException) {
    return CacheFailure.fromException(exception);
  } else if (exception is UnknownException) {
    return UnknownFailure.fromException(exception);
  }
  return UnknownFailure("An unknown error has occurred", 500);
}
